### PR TITLE
Add Mission Control 2

### PR DESCRIPTION
**Pull request by jerith**

### CODE REVIEW

**Comment by jerith** [comment]
This is a fragile hack that uses a shell script to send an app definition to Marathon and a separate `consul-template` template to expose it on port 7001 of the host that has the `seed_stack::mc2` class applied to it.

The two major changes that need to happen are switching to something like https://github.com/praekelt/marathon-sync for the marathon group management and exposing MC2 somewhere better than port 7001.

File: files/mc2.marathon.json [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+            {"key": "volume-driver", "value": "xylem"},
+            {"key": "volume", "value": "seed-infra-mc2:/data"}
+          ]
+        }
+      },
+      "env": {
+        "MESOS_MARATHON_HOST": "http://marathon.service.consul:8080",
+        "SECRET_KEY": "seekrit",
+        "PROJECT_ROOT": "/deploy/",
+        "DATABASE_URL": "sqlite:////data/mc2.db",
```
Comment (calston) [inline files/mc2.marathon.json]: Really? Sqlite?

Comment (jerith) [inline files/mc2.marathon.json]: We don't have xylem postgres in here yet. This is a hacky prototype. :-)

File: files/nginx-mc2.ctmpl [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+#       to this file.
+#
+#       This expects a key to exist under "consular/<app_id>/domain"
+#       in Consul's KV store. This is used to map a domain to a service in
+#       Nginx. If the key does not exist the service isn't added to the
+#       list of services in the Nginx config.
+
+{{range services}}{{$labels := ls (print "consular/" .Name) | explode }}{{if eq .Name "seed-infra-mc2-mc2"}}
+server {
+  listen 7001;
```
Comment (JayH5) [inline files/nginx-mc2.ctmpl]: Is there a reason we have to have MC2 on port 7001?

File: manifests/mc2.pp [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+#
+# Runs a Mission Control container in Marathon.
+#
+class seed_stack::mc2 {
+
+  unless defined(Package['curl']) {
+    package { 'curl': ensure => 'installed' }
+  }
+
+  file { '/usr/local/sbin/manage-marathon-group.sh':
```
Comment (JayH5) [inline manifests/mc2.pp]: Why `sbin`?

**Comment by jerith** [comment]
The questions I have around this:
- How do we want to expose this to the outside world? I'm currently sticking it on port 7001 because that's what the docker command I cribbed from did.
- How do we want to manage the database? I chose sqlite on a xylem volume because it was easy to get running. I think that's fine for a prototype thing, but not really for production.
- How do we want to keep the group definition in Marathon up to date? A shell script hitting the API is really fragile, especially if the cluster isn't guaranteed to be set up yet when puppet runs it.

File: manifests/mc2.pp (hunk 3 above)
Comment (jerith) [inline manifests/mc2.pp]: No real reason. I probably just saw something else that put a script in `/usr/local/sbin` and did the same.

File: files/nginx-mc2.ctmpl (hunk 2 above)
Comment (jerith) [inline files/nginx-mc2.ctmpl]: Nope. I needed to expose it somewhere so I could test that it was working, and the docker run command I used for reference put it on 7001.

File: manifests/mc2.pp (hunk 3 above)
Comment (JayH5) [inline manifests/mc2.pp]: I thought `sbin` was generally for things that needed to run as root.

File: files/nginx-mc2.ctmpl (hunk 2 above)
Comment (JayH5) [inline files/nginx-mc2.ctmpl]: I think we can just expose MC2 from its URL the same way we do anything else.

File: manifests/mc2.pp (hunk 3 above)
Comment (calston) [inline manifests/mc2.pp]: sbin is for binaries which should only be accessible by root

File: manifests/mc2.pp [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+# Runs a Mission Control container in Marathon.
+#
+class seed_stack::mc2(
+  $infr_domain,
+  $hub_domain,
+  $marathon_host = 'marathon.service.consul',
+) {
+  file { '/etc/marathon-apps': ensure => 'directory' }
+  ->
+  file { '/etc/marathon-apps/mc2.marathon.json':
```
Comment (JayH5) [inline manifests/mc2.pp:12]: I'm wondering if `puppet-xylem` should actually handle the placement of these files... but it's mostly semantics. Something like:

``` puppet
xylem::marathon_sync::group { 'mc2':
  content => template('seed_stack/mc2.marathon.json.erb')
}
```

I donno, just a thought..

**Comment by JayH5** [comment]
:+1: ... assuming you've tested this?

File: manifests/mc2.pp (hunk 4 above)
Comment (jerith) [inline manifests/mc2.pp:12]: I don't think xylem should mandate the location of these files. They might live on a shared volume or something.

**Comment by jerith** [comment]
I have. Thanks. :-)

**Comment by calston** [comment]
:+1:

**Pull request merged**